### PR TITLE
Add logic for clusterconfigmap and podcidr resources

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/roadmap#320

Adds clusterconfigmap resource and internal podcidr resource.

```sh
D 05/18 09:29:07 ross/ross clusterconfigmap.GetCurrentState finding cluster config maps in namespace `ross` | cluster-apps-operator/service/controller/resource/clusterconfigmap/current.go:24 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.GetCurrentState found 0 config maps in namespace `ross` | cluster-apps-operator/service/controller/resource/clusterconfigmap/current.go:39 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computing ConfigMaps to create | operatorkit/v4/pkg/resource/k8s/configmapresource/create.go:47 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computed 1 ConfigMaps to create | operatorkit/v4/pkg/resource/k8s/configmapresource/create.go:55 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computing ConfigMaps to delete | operatorkit/v4/pkg/resource/k8s/configmapresource/delete.go:55 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computed 0 ConfigMaps to delete | operatorkit/v4/pkg/resource/k8s/configmapresource/delete.go:63 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computing ConfigMaps to update | operatorkit/v4/pkg/resource/k8s/configmapresource/update.go:43 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.NewUpdatePatch computed 0 ConfigMaps to update | operatorkit/v4/pkg/resource/k8s/configmapresource/update.go:54 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.ApplyCreateChange creating ConfigMap `ross-cluster-values` in namespace `ross` | operatorkit/v4/pkg/resource/k8s/configmapresource/create.go:20 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
D 05/18 09:29:07 ross/ross clusterconfigmap.ApplyCreateChange created ConfigMap `ross-cluster-values` in namespace `ross` | operatorkit/v4/pkg/resource/k8s/configmapresource/create.go:28 | controller=cluster-apps-operator-cluster-controller | event=update | loop=0 | version=245238279
```

```yaml
$ k -n ross get cm ross-cluster-values -o yaml | k neat

apiVersion: v1
data:
  values: |
    baseDomain: ross.k8s.ghost.westeurope.azure.gigantic.io
    cluster:
      calico:
        CIDR: 10.1.0.0/16/16
      kubernetes:
        API:
          clusterIPRange: 172.31.0.0/16
        DNS:
          IP: 172.31.0.10
    clusterDNSIP: 172.31.0.10
    clusterID: ross
kind: ConfigMap
metadata:
  annotations:
    giantswarm.io/notes: DO NOT EDIT. Values managed by cluster-apps-operator.
  labels:
    giantswarm.io/cluster: ross
    giantswarm.io/managed-by: cluster-apps-operator
  name: ross-cluster-values
  namespace: ross
```

Base domain is expected because config is not set up yet.

### CODE REVIEW

File: flag/flag.go [hunk 1]
```diff
@@ -4,13 +4,11 @@ import (
...
 
 	"github.com/giantswarm/cluster-apps-operator/flag/service"
-	"github.com/giantswarm/cluster-apps-operator/flag/workload"
 )
 
 // Flag provides data structure for service command line flags.
 type Flag struct {
-	Service  service.Service
-	Workload workload.Workload
+	Service service.Service
```
Comment (rossf7) [inline flag/flag.go:11]: I moved the workload flags under service as that is our usual approach.

File: helm/cluster-apps-operator/values.yaml [hunk 2]
```diff
@@ -8,7 +10,7 @@ image:
 
 kubernetes:
   api:
-    clusterAPIRange: 172.31.0.0/16
+    clusterIPRange: 172.31.0.0/16
```
Comment (rossf7) [inline helm/cluster-apps-operator/values.yaml:13]: Just fixing this.

File: service/internal/podcidr/pod_cidr.go [hunk 3]
```diff
@@ -38,6 +45,49 @@ func New(c Config) (*PodCIDR, error) {
...
+		return "", microerror.Mask(err)
+	}
+
+	cl, err := p.lookupCluster(ctx, cr)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	var podCIDR string
+	if cl.Spec.ClusterConfiguration == nil || cl.Spec.ClusterConfiguration.Networking.PodSubnet == "" {
```
Comment (rossf7) [inline service/internal/podcidr/pod_cidr.go:58]: ClusterConfiguration is a pointer so added this nil check to fix a crash.

**Comment by rossf7** [comment]
Tested on ginger.

File: service/internal/podcidr/pod_cidr.go (hunk 3 above)
Comment (tomahawk28) [reply service/internal/podcidr/pod_cidr.go:58]: isn't that what you want to?
```suggestion
	if cl.Spec.ClusterConfiguration =! nil && cl.Spec.ClusterConfiguration.Networking.PodSubnet == "" {
```

File: service/controller/resource/clusterconfigmap/current.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) ([]*corev1.ConfigMap, error) {
+	cr, err := key.ToCluster(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var configMaps []*corev1.ConfigMap
+	{
+		r.logger.Debugf(ctx, "finding cluster config maps in namespace %#q", key.ClusterID(&cr))
```
Comment (tomahawk28) [inline service/controller/resource/clusterconfigmap/current.go:24]: I'm not sure how we logged this before, but better to concatenate two words.
```suggestion
		r.logger.Debugf(ctx, "finding cluster configmaps in namespace %#q", key.ClusterID(&cr))
```

File: service/controller/resource/clusterconfigmap/current.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		list, err := r.k8sClient.CoreV1().ConfigMaps(key.ClusterID(&cr)).List(ctx, lo)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		for _, item := range list.Items {
+			configMaps = append(configMaps, item.DeepCopy())
+		}
+
+		r.logger.Debugf(ctx, "found %d config maps in namespace %#q", len(configMaps), key.ClusterID(&cr))
```
Comment (tomahawk28) [inline service/controller/resource/clusterconfigmap/current.go:39]:
```suggestion
		r.logger.Debugf(ctx, "found %d configmaps in namespace %#q", len(configMaps), key.ClusterID(&cr))
```

**Review by tomahawk28 — COMMENTED** [review]
Looks great; just a few suggestions.

File: service/controller/resource/clusterconfigmap/current.go (hunk 4 above)
Comment (rossf7) [reply service/controller/resource/clusterconfigmap/current.go]: Yes, that's better. Thx!

File: service/internal/podcidr/pod_cidr.go (hunk 3 above)
Comment (rossf7) [reply service/internal/podcidr/pod_cidr.go:58]: The cluster-operator code doesn't have the nil check but the operator crashed in testing. As this block in the CR was empty.

This treats a nil ClusterConfiguration the same as the string being empty which I think is correct.

**Review by rossf7 — COMMENTED** [review]
@tomahawk28 PTAL

**Review by tomahawk28 — APPROVED** [review]

**Pull request merged**